### DISCUSSION TITLE
Performance of garbage collection in HDF5 files - cgio_compress_file

### DISCUSSION THREAD

**Discussion by mfolusiak** [CGNS/CGNS#348] [CGNStalk]
I am using in my project a garbage collection function from the CGNS library with (quite confusing) name `cgio_compress_file()`. The function rewrites the CGNS file on disk and orphaned nodes are omitted, thus the file size is reduced. This function once has been documented in the user manual and is no longer there in version >4. I am wondering if this is for a reason of poor performance or different design concept/ good practice I am not aware of.
I am using CGIO in all my function calls so the performance of file operations is typically OK. However, yesterday I performed some experiments on some CGNS files that have large number of user-defined nodes written. I noticed that the time spent in  `cgio_compress_file()` is proportional to `N^2`, where N is number of nodes. I think it is quite an issue for large CGNS files and I would expect the relation to be linear, not power.

![image](https://user-images.githubusercontent.com/6044635/176441727-b9cba790-efc5-47cd-8d2f-f572229df06a.png)

I would like to ask the community if you have similar experience and what should be the best practice for the garbage collection in HDF5 files.

**Comment by mennodeij** [comment]
<strike>As far as I know and understand, compression has been removed in 4.x because parallel I/O does not support it. In parallel I/O you need to find an offset in the file to start reading from, and if that is changed due to compression, you will read wrong data or get I/O error.

**Comment by brtnfld** [comment]
@mennodeij different kind of compression. BTW, if you are interested in trying proper compression, it's being developed in the hdf5_compression branch.

I think this is just an oversight in the documentation, and there are no plans to remove cgio_compress_file. I've not seen or heard of performance issues with the function, but then again, I'm not aware of anyone doing a study of it.  

Can you profile (maybe using gprof or some other tool) where the time is spent for many datasets cases? My guess would be H5Literate.

Which version of HDF5 are you using? HDF5 1.12 has some performance improvements in H5Oget_info_by_name3.

Deleting a node calls H5Literate (multiple times) for deleting nodes, which is not a cheap operation. Furthermore, the callback for H5Literate also calls H5Literate, so I would not expect the performance to be linear. The more child nodes there are, the more iterations need to be done.

Basically, if you delete a link, it has to check if it had other links that needed to be deleted, and so forth down the rabbit hole of links.

I think it would be O(infinity) as the upper limit. The only way to get O(n) is if there are no cycles in the callback to H5Literate.
But as you delete nodes, say at a constant rate, I guess it could reduce to O(n^2) or O(n).

**Reply by mennodeij** [reply]
Ah, ok, thanks for clearing that up.

**Comment by brtnfld** [comment]
I will have to do some more digging because I don't see anywhere cgio_compress_file is calling any node delete functions. However, I see where it checks if a link exists, so it probably is not an H5Literate issue.

Are you talking about the tool cgnscompress? Unfortunately, I don't see anywhere in the past documentation where cgio_compress_file was documented. Do you have a link?

By what amount are the files reduced? I would be surprised if just deleting an empty link would save that much space in the file. Have you verified the resulting files are valid? I don't think there are any regression tests for cgio_compress_file (or cgnscompress).

I would be curious to see the difference between the two files if you can do h5dump for the two files and do a diff between the results.

If you can make an example file available I could take a look.

**Reply by mfolusiak** [reply]
@brtnfld thank you
You are right wrt the documentation of the `cgio_compress_file` I cannot find this in the older documentation too, so I must have mixed things up, so far I just believed that I have learned about this function from the documentation, not the source code.
Anyway, I will reduce my example and provide you with link to files / or the source code to generate them tomorrow.
I tested the difference with cgnsdiff, they are the same structure-wise, the reduction in size was minimal (but nonzero) cause no node deletion happened in this case after wiring the file.

**Comment by mfolusiak** [comment]
Hi,
please find below an example code that can be used to reproduce the behavior. The program takes only one argument - integer number of identical nodes to create. It prints the elapsed time in milliseconds. Here are the performance measurements (as you can see the input is doubled every time and the elapsed time gets ~ quadrupled):
```bash
> ./cgns_compression 600
Elapsed time 1.347000 seconds 
> ./cgns_compression 1200
Elapsed time 5.041000 seconds 
> ./cgns_compression 2400
Elapsed time 19.923000 seconds 
> ./cgns_compression 4800
Elapsed time 90.167000 seconds 
```
Code:
```cpp
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <chrono>

#include "cgns_io.h"

/* --- Data to be stored in database */
float DATA[3][4] = {{1.1,2.1,3.1,4.1},
                    {1.2,2.2,3.2,4.2},
                    {1.3,2.3,3.3,4.3}
                   };
cgsize_t DATA_DIM[2] = {4,3};

void create_numbered_node_with_data(int cgio_num, double parent_id, int number)
{
   double node_id;
   char label[CGIO_MAX_LABEL_LENGTH+1];
   sprintf(label, "node%d", number);
   cgio_create_node(cgio_num,parent_id,label,&node_id);
   cgio_set_label(cgio_num,node_id,"UserDefinedData_t");

   cgio_set_dimensions(cgio_num,node_id,"R4",2,DATA_DIM);
   cgio_write_all_data(cgio_num,node_id,DATA);
}

using namespace std::chrono;

int main(int argc, char** argv)
{
   char data_type[CGIO_MAX_DATATYPE_LENGTH+1];
   int cgio_num;
   double root_id;
   const char *fn="file.cgns";

   // the only argument is number of nodes to create
   auto cnt=atoi(argv[1]);

   cgio_open_file(fn,CGIO_MODE_WRITE,CGIO_FILE_HDF5,&cgio_num);
   cgio_get_root_id(cgio_num,&root_id);

   for (auto i=0; i<cnt ; ++i)
      create_numbered_node_with_data(cgio_num, root_id, i);

   auto start = high_resolution_clock::now();
   cgio_compress_file(cgio_num,fn);
   auto end = high_resolution_clock::now();
   auto duration = duration_cast<milliseconds>(end - start);
   printf("Elapsed time %f seconds \n",duration.count()/1000.);
   cgio_close_file(cgio_num);

   return 0;
}
```

**Comment by brtnfld** [comment]
cgio_compress_file does not do anything for this use case, the file is basically the same after compression.
The majority of the time is spent in a skip list search in HDF5, I'm not sure how to speed that up except trying to minimize the number of times it is called.

**Reply by mfolusiak** [reply]
> cgio_compress_file does not do anything for this use case, the file is basically the same after compression.
> The majority of the time is spent in a skip list search in HDF5, I'm not sure how to speed that up except trying to minimize the number of times it is called.

I added to my exemplary code part where I open compressed `file.cgns` for reading and the new `file2.cgns` where I write all content read from the compressed file. The example is quite naive - not recursive, doesn't follow links, etc. However it IMO demonstrates that compression of complexity of N instead N^2 is possible by just iterating and re-writing nodes.
Here are the results:
```bash
> ./cgns_compression 600
Elapsed time 1.602000 seconds (cgio_compress_file)
Elapsed time 0.230000 seconds (re-write)
> ./cgns_compression 1200
Elapsed time 6.166000 seconds (cgio_compress_file)
Elapsed time 0.473000 seconds (re-write)
> ./cgns_compression 2400
Elapsed time 25.303000 seconds (cgio_compress_file)
Elapsed time 1.021000 seconds (re-write)
> ./cgns_compression 4800
Elapsed time 104.245000 seconds (cgio_compress_file)
Elapsed time 2.630000 seconds (re-write)

```
Here is the code:
```c++
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <chrono>

#include "cgns_io.h"

/* --- Data to be stored in database */
float DATA[3][4] = {{1.1,2.1,3.1,4.1},
                    {1.2,2.2,3.2,4.2},
                    {1.3,2.3,3.3,4.3}
                   };
cgsize_t DATA_DIM[2] = {4,3};

void create_numbered_node_with_data(int cgio_num, double parent_id, int number)
{
   double node_id;
   char label[CGIO_MAX_LABEL_LENGTH+1];
   sprintf(label, "node%d", number);
   cgio_create_node(cgio_num,parent_id,label,&node_id);
   cgio_set_label(cgio_num,node_id,"UserDefinedData_t");

   cgio_set_dimensions(cgio_num,node_id,"R4",2,DATA_DIM);
   cgio_write_all_data(cgio_num,node_id,DATA);
}

using namespace std::chrono;

int main(int argc, char** argv)
{
   char data_type[CGIO_MAX_DATATYPE_LENGTH+1];
   int cgio_num;
   double root_id;
   const char *fn="file.cgns";

   // the only argument is number of nodes to create
   auto cnt=atoi(argv[1]);

   cgio_open_file(fn,CGIO_MODE_WRITE,CGIO_FILE_HDF5,&cgio_num);
   cgio_get_root_id(cgio_num,&root_id);

   for (auto i=0; i<cnt ; ++i)
      create_numbered_node_with_data(cgio_num, root_id, i);

   auto start = high_resolution_clock::now();
   cgio_compress_file(cgio_num,fn);
   auto end = high_resolution_clock::now();
   auto duration = duration_cast<milliseconds>(end - start);
   printf("Elapsed time %f seconds (cgio_compress_file)\n",duration.count()/1000.);
   cgio_close_file(cgio_num);

   start = high_resolution_clock::now();
   int cgio_num2;
   const char *fn2="file2.cgns";
   double root_id2;
   cgio_open_file(fn,CGIO_MODE_READ,CGIO_FILE_HDF5,&cgio_num);
   cgio_open_file(fn2,CGIO_MODE_WRITE,CGIO_FILE_HDF5,&cgio_num2);
   cgio_get_root_id(cgio_num,&root_id);
   cgio_get_root_id(cgio_num2,&root_id2);

   /*
    * open new file, create identical nodes with the coppied content
    */
   int num_ret;
   cgio_number_children(cgio_num, root_id, &cnt);
   double *child_ids = new double[cnt];
   cgio_children_ids(cgio_num, root_id, 1, cnt, &num_ret, child_ids);
   for (auto i=0; i<cnt; ++i)
   {
       char label[CGIO_MAX_LABEL_LENGTH+1];
       double node_id, node_id2;
       node_id=child_ids[i];
       cgio_get_name(cgio_num,node_id,label);
       cgio_create_node(cgio_num2,root_id2,label,&node_id2);
       cgio_get_label(cgio_num,node_id,label);
       cgio_set_label(cgio_num2,node_id2,label);

       char data_type[3];
       int ndims;
       cgsize_t dims[2];
       cgio_get_data_type(cgio_num, node_id, data_type);
       cgio_get_dimensions(cgio_num, node_id, &ndims, dims);
       cgio_set_dimensions(cgio_num2,node_id2,data_type,ndims,dims);
       cgio_read_all_data_type(cgio_num,node_id,data_type,DATA);
       cgio_write_all_data(cgio_num2,node_id2,DATA);
}

   cgio_close_file(cgio_num);
   cgio_close_file(cgio_num2);
   end = high_resolution_clock::now();
   duration = duration_cast<milliseconds>(end - start);
   printf("Elapsed time %f seconds (re-write)\n",duration.count()/1000.);

   return 0;
}
```

**Reply by mfolusiak** [reply]
Hi @brtnfld ,
do you think the behavior described above is expected. Should I raise Bug or Improvement in JIRA?
Best,
Michal

**Comment by brtnfld** [comment]
I think it would be good to document the issue in JIRA and mark it as an improvement.
